Break down tests having multiple checks and xfail decorated

matteo-pallini opened the pull request:
Closes #4425 

Xfail is going to care only about the first failed assert statement and
ignore all the other ones in the same test.

Sometimes xfail is being used to decorate a test or class function calling
several other sub-functions. This makes it hard to monitor what is happening
for the sub-fuctions that are not failing (given that the failure will take
priority and mark the test function as xfailed). By breaking down each
function into individual sub-functions it's easier to monitor individual
tests behavior


**Thank your for opening a PR!**

Depending on what your PR does, here are a few things you might want to address in the description:
+ [x] important background, or details about the implementation
+ [x] [linting/style checks have been run](https://github.com/pymc-devs/pymc3/wiki/PyMC3-Python-Code-Style)

matteo-pallini committed:
Break down tests having multiple checks and xfail decorated

Xfail is going to care only about the first failed assert statement and
ignore all the other ones in the same test.

Sometimes xfail is being used to decorate a test or class function calling
several other sub-functions. This makes it hard to monitor what is happening
for the sub-fuctions that are not failing (given that the failure will take
priority and mark the test function as xfailed). By breaking down each
function into individual sub-functions it's easier to monitor individual
tests behavior

matteo-pallini committed:
Remove xfail which seems unnecessary

Also add comment regarding test with unpredictable outcome

matteo-pallini committed:
Break down test with multiple sub-checks and xfail where appropriate

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -889,14 +889,18 @@ def test_half_flat(self):
         )
+
+    @pytest.mark.xfail(condition=(aesara.config.floatX == "float32"), reason="Fails on float32")
+    # this test doesn't always result in the same outcome (pass/not pass)
```
How did you test this in float32? Locally or in here?

To make it reproducible you can temporarily set `n_samples=-1` in `check_logp`/`check_logcdf`. This will force it to run all permutations of the test instead of a limited subset.

Just want to confirm this is not an issue we have overlooked since the test was not marked as xfail before. Maybe it will be enough to lower the float32 precision like is being done for check_logp already.

matteo-pallini replied:
I ran all the tests locally by setting floatX to float32

After having set `n_samples=-1` it returns the same result (failure) consistently, thanks. I am going to remove the comment

ricardoV94 replied:
This one should be scrutinized a bit more because it was not marked as xfail before (right?). 

Do you feel capable of identifying which values are failing (input -> output -> expected_output in float32)? You might need to temporarily set a try/except in check_logcdf to not interrupt the test prematurely. Let me know if you need help with this.

ricardoV94 replied:
We can wait and see if it comes as an xfail here to be sure there is actually a problem, before delving deeper.

ricardoV94 replied:
Indeed it XFAILED: https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

```
pymc3/tests/test_distributions.py::TestMatchesScipy::test_normal_logcdf XFAIL [  2%]
```

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1244,6 +1250,7 @@ def test_fun(value, mu, sigma):
         condition=(aesara.config.floatX == "float32"),
         reason="Fails on float32 due to numerical issues",
     )
+    # this test doesn't always result in the same outcome fail/success
```
As mentioned above, that is probably due to subsampling of the test combinations, since the default is `n_samples=100`. I would either remove this comment or substitute it by a more informative one that explains why this is not a deterministic failure. 

Same for the other repeated comments

matteo-pallini replied:
Same as above. Once added `n_samples=-1` the result is consistent. Removing the comment

ricardoV94 replied:
Xfailed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1371,24 +1389,36 @@ def test_binomial(self):
         )
-        self.check_logp(
+
+    @pytest.mark.xfail(
```
This one does not require scipy > 1.4.0 right?

Since it doesn't fail you can also combine it with the test_beta_binomial_discrete_logcdf below.

matteo-pallini replied:
You are right, thanks for spotting

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1475,27 +1505,39 @@ def test_constantdist(self):
+        condition=(aesara.config.floatX == "float32"),
+        reason="Fails on float32 due to inf issues",
+    )
+    # this test doesn't always result in the same outcome (pass/not pass)
```
This one should always pass or fail, since it has no random subsampling... Are you sure?

matteo-pallini replied:
I haven't looked into the test logic but it seems to return both pass and fails on my machine. I will investigate

![Screenshot from 2021-03-03 10-29-26](https://user-images.githubusercontent.com/11328307/109792950-f7def280-7c0b-11eb-9d03-0c22935c123d.png)

ricardoV94 replied:
You are right. I see now that `checkd` calls `check_int_to_1`, which has an `n_samples=10` in it. So the same source of randomness as the other tests. Thanks for checking

ricardoV94 replied:
We should also temporarily change this one inplace for n_samples=-1, in case we removed any xfail over a checkd distribution. Check it's fine and then put it back to 10. The relevant line is here: https://github.com/pymc-devs/pymc3/blob/ea1b03811f7d58a38c09a56a61144294f3732d71/pymc3/tests/test_distributions.py#L775

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1475,27 +1505,39 @@ def test_constantdist(self):
+    def test_zeroinflatednegativebinomial_logcdf(self):
         self.check_selfconsistency_discrete_logcdf(
             ZeroInflatedNegativeBinomial,
             Nat,
```
This one is important we temporarily set n_samples to -1, as we recently changed it to 10 for speed reasons.

The test will take ages to run but we will know for sure that it does not need an xfail. We don't want to introduce any flaky test down the road.

Actually we should do this for all removed xfails

ricardoV94 replied:
Passed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1504,13 +1546,14 @@ def test_zeroinflatednegativebinomial(self):
         )
+
+    def test_zeroinflatedbinomial_logcdf(self):
         self.check_selfconsistency_discrete_logcdf(
```
Same here we are only testing n_samples=10, and we need to run once with -1 to make sure it does not need an xfail

ricardoV94 replied:
Passed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_models_utils.py:
```diff
@@ -51,6 +50,8 @@ def test_dict_input(self):
         m, l = utils.any_to_tensor_and_labels(self.data.to_dict("list"))
         self.assertMatrixLabels(m, l, mt=self.data[l].values, lt=l)
 
+    @pytest.mark.xfail
```
Can we dig the reason for this one?

matteo-pallini replied:
I think that `pymc3.glm.utils.any_to_tensor_and_labels` doesn't support converting data structures containing TensorType (mostly because `theano.tensor.as_tensor_variable` called on line 132 doesn't). So, in this case `inp` will cause an error to be raised

I haven't been able to understand what conditions made this test pass. I guess that if we need `pymc3.glm.utils.any_to_tensor_and_labels` to be able to handle data structure containing tensor objects we should probably change that

ricardoV94 replied:
I tracked it down to this commit: https://github.com/pymc-devs/pymc3/commit/4f67d639f984195f81b03cbfc19de64507f9f1f1

@rpgoldman could you chime in? What is the source of this failure? It would be helpful to have a more clear reason specified in the xfail.

michaelosthege replied:
The xfail was there before, so this shouldn't block us from moving forward with this PR.

ricardoV94 reviewed (changes requested):
I checked it now. Great work disentangling things! 

I left a couple of comments that need to be addressed.

Importantly, we have to make sure that all the removed xfails around `check_logp` and `check_logcdf` are warranted. For this we have to temporarily set `n_samples=-1` in the `check_logp` and `check_logcdf` methods to force pytest to test all combinations of parameter values. After confirming the tests do not fail on float32, we can revert back to the original n_samples.

matteo-pallini committed:
Remove Scipy Xfail from tests not using sp.betabinom

matteo-pallini committed:
Add back xfail to potentially failing test when float32 is used

matteo-pallini committed:
Temporatily add n_samples=-1 to check tests behavior for all possible parameters

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1371,28 +1398,38 @@ def test_binomial(self):
+        self.checkd(
             BetaBinomial,
             Nat,
             {"alpha": Rplus, "beta": Rplus, "n": NatSmall},
```
This one should be tested with n_samples=-1, but you have to change it in `check_int_to_1`. Maybe let's add the optional argument `n_samples=10` to `check_int_to_1`, and then you can temporarily call this test via:

```python
    def test_beta_binomial(self):
        self.checkd(
            BetaBinomial,
            Nat,
            {"alpha": Rplus, "beta": Rplus, "n": NatSmall},
            checks=(lambda *args: self.check_int_to_1(*args, n_samples=-1),
       )

ricardoV94 replied:
This one failed in here: https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

In checkd:
```E           AssertionError: 
E           Arrays are not almost equal to 7 decimals
E           {'alpha': array(0.1, dtype=float32), 'beta': array(0.01, dtype=float32), 'n': array(5), 'value': array(0, dtype=int16)}
E            ACTUAL: 0.9999997497252575
E            DESIRED: 1
```

matteo-pallini replied:
Xfail added

matteo-pallini replied:
Xfailed as expected

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1371,28 +1398,38 @@ def test_binomial(self):
+        self.check_selfconsistency_discrete_logcdf(
             BetaBinomial,
             Nat,
             {"alpha": Rplus, "beta": Rplus, "n": NatSmall},
```
Also needs n_samples=-1

matteo-pallini replied:
changed

matteo-pallini replied:
passed as expected

ricardoV94 commented:
Let's keep an eye on this one: https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

It will show us whether the removed xfails were safe, and whether the added one for the normal logcdf was needed or not. After this we can go ahead an test the removed xfails that are still not getting n_samples=-1.

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1475,47 +1512,60 @@ def test_constantdist(self):
             ZeroInflatedBinomial,
             Nat,
             {"n": NatSmall, "p": Unit, "psi": Unit},
         )
```
This needs the same trick as the `beta_binomial::check_d` mentioned above

matteo-pallini replied:
added

matteo-pallini replied:
passed as expected

ricardoV94 commented:
This one on float64 is new: https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551515

```
FAILED pymc3/tests/test_distributions.py::TestMatchesScipy::test_moyal_logp
```

```
E           AssertionError: 
E           Arrays are not almost equal to 6 decimals
E           {'mu': array(2.1), 'sigma': array(0.5), 'value': array(-2.1)}
E           x and y -inf location mismatch:
E            x: array(-2219.559165)
E            y: array(-inf)
```
Looks like a scipy numerical issue:
```
st.moyal(2.1, 0.5).logpdf(-2.1)
# -inf
```

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -889,19 +889,25 @@ def test_half_flat(self):
             {"mu": R, "sigma": Rplus},
             lambda value, mu, sigma: sp.norm.logpdf(value, mu, sigma),
             decimal=select_by_precision(float64=6, float32=1),
+            n_samples=-1,
```
Passed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -941,23 +947,27 @@ def test_chi_squared(self):
             {"mu": Rplus, "alpha": Rplus},
             lambda value, mu, alpha: sp.invgauss.logpdf(value, mu=mu, loc=alpha),
             decimal=select_by_precision(float64=6, float32=1),
+            n_samples=-1,
```
Passed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -941,23 +947,27 @@ def test_chi_squared(self):
             Rplus,
             {"mu": Rplus, "alpha": Rplus},
             lambda value, mu, alpha: sp.invgauss.logcdf(value, mu=mu, loc=alpha),
+            n_samples=-1,
```
Xfailed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1254,28 +1265,35 @@ def test_gamma_logcdf(self):
             Rplus,
             {"alpha": Rplus, "beta": Rplus},
             lambda value, alpha, beta: sp.invgamma.logpdf(value, alpha, scale=beta),
+            n_samples=-1,
```
Passed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1254,28 +1265,35 @@ def test_gamma_logcdf(self):
             {"alpha": Rplus, "beta": Rplus},
             lambda value, alpha, beta: sp.invgamma.logcdf(value, alpha, scale=beta),
             skip_paramdomain_outside_edge_test=True,
+            n_samples=-1,
```
Xfailed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1313,18 +1331,27 @@ def test_pareto(self):
             Rplus,
             {"alpha": Rplusbig, "beta": Rplusbig},
             lambda value, alpha, beta: sp.exponweib.logpdf(value, 1, alpha, scale=beta),
+            n_samples=-1,
```
Xfailed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1313,18 +1331,27 @@ def test_pareto(self):
             Rplus,
             {"alpha": Rplusbig, "beta": Rplusbig},
             lambda value, alpha, beta: sp.exponweib.logcdf(value, 1, alpha, scale=beta),
+            n_samples=-1,
```
Xfailed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1371,28 +1398,38 @@ def test_binomial(self):
             Nat,
             {"alpha": Rplus, "beta": Rplus, "n": NatSmall},
+            lambda value, alpha, beta, n: sp.betabinom.logcdf(value, a=alpha, b=beta, n=n),
+            n_samples=-1,
```
Xfailed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1371,28 +1398,38 @@ def test_binomial(self):
             Nat,
             {"alpha": Rplus, "beta": Rplus, "n": NatSmall},
+            lambda value, alpha, beta, n: sp.betabinom.logpmf(value, a=alpha, b=beta, n=n),
+            n_samples=-1,
```
Xfailed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

Due to old scipy

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1475,47 +1512,60 @@ def test_constantdist(self):
             ZeroInflatedPoisson,
             Nat,
             {"theta": Rplus, "psi": Unit},
+            n_samples=-1,
```
Passed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -2280,18 +2330,23 @@ def test_rice(self):
             R,
             {"mu": R, "sigma": Rplusbig},
             lambda value, mu, sigma: floatX(sp.moyal.logpdf(value, mu, sigma)),
+            n_samples=-1,
```
Xfailed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -2280,18 +2330,23 @@ def test_rice(self):
             R,
             {"mu": R, "sigma": Rplusbig},
             lambda value, mu, sigma: floatX(sp.moyal.logcdf(value, mu, sigma)),
+            n_samples=-1,
```
Xfailed as expected in https://github.com/pymc-devs/pymc3/pull/4497/checks?check_run_id=2021551266

ricardoV94 reviewed (changes requested):
I went ahead and checked what Passed/Failed as expected. Everywhere I wrote "Passed/XPassed" as expected you can revert the n_samples=-1, and mark the conversation as resolved. 

Almost there!!!!

ricardoV94 commented on pymc3/tests/test_distributions.py:
```diff
@@ -1371,24 +1389,36 @@ def test_binomial(self):
-        self.check_logcdf(
+
+    @pytest.mark.xfail(
+        condition=(SCIPY_VERSION < parse("1.4.0")), reason="betabinom is new in Scipy 1.4.0"
```
Would this be better as a `@pytest.mark.skipif`?

matteo-pallini replied:
Possibly yes. Given that there should be no chance for the behavior of `betabinom` to change, given that it won't exist there is no point in monitoring the outcome of these tests, so it may be worth skipping them. I would expect that skipif is faster than xfail

matteo-pallini commented on pymc3/tests/test_distributions.py:
```diff
@@ -1261,6 +1265,7 @@ def test_gamma_logcdf(self):
             {"alpha": Rplusbig, "beta": Rplusbig},
             lambda value, alpha, beta: sp.gamma.logcdf(value, alpha, scale=1.0 / beta),
             skip_paramdomain_outside_edge_test=True,
+            n_samples=-1,
```
passed as expected

ricardoV94 replied:
You mean xfailed as expected?

matteo-pallini replied:
yeah, sorry

ricardoV94 replied:
Just making sure we didn't miss anything :D

matteo-pallini commented on pymc3/tests/test_distributions.py:
```diff
@@ -906,6 +907,7 @@ def test_normal_logcdf(self):
             R,
             {"mu": R, "sigma": Rplus},
             lambda value, mu, sigma: sp.norm.logcdf(value, mu, sigma),
+            n_samples=-1,
```
xfailed as expected

ricardoV94 commented:
Great work. Do you want to remove the unnecessary n_samples changes and comments, and check those that we skipped on the first run?

matteo-pallini commented on pymc3/tests/test_distributions.py:
```diff
@@ -1553,7 +1565,7 @@ def test_zeroinflatedbinomial_logcdf(self):
             Nat,
             {"n": NatSmall, "p": Unit, "psi": Unit},
-            n_samples=10,
+            n_samples=-1,
```
passed as expected

matteo-pallini committed:
Remove temporary n_samples and a couple more to do some quick tests

matteo-pallini committed:
Remove unnecessary comments

matteo-pallini committed:
Replace Xfail with skipif when using Scipy version as condition

Given that betabinom won't exist in previous versions, there should be
no chance for it's behavior to change, given that it won't exist there
is no point in monitoring the outcome of these tests, so it may be worth
skipping them. Skipping them should be faster than running them and
catching the failures

ricardoV94 commented:
Listing here the issues revealed by this PR that need some thinking:
- [x] Moyal check_logp fails on 64bit, because the scipy version is more numerically unstable. We shouldn't simply mark it as xfail on 64 bit as well, because then we loose coverage of this distribution logp altogether. Might need to change the test domain... or hand pick a set of values and use pytest.mark.parametrize. **Fixed by tweaking testing domain slightly**
- [x] Normal check_logcdf fails on 32bit. Reasons still unclear...  **Fixed by changing precision of evaluation on 32bit**
- [ ] `glm.utils.any_to_tensor_and_labels fails` when given a dictionary with `as_tensor_variable` objects. Is this okay, is it not?

Anything else that was found to have problems for the first time by this PR @drabbit17

matteo-pallini commented:
Possibly the fact that `pymc3.glm.utils.any_to_tensor_and_labels` cannot handle data structures (or at least dict) containing TensorType objects? Although, I am not sure whether it's meant to handle these in the first place

matteo-pallini commented on pymc3/tests/test_distributions.py:
```diff
@@ -1363,36 +1378,47 @@ def test_binomial(self):
             Nat,
             {"n": NatSmall, "p": Unit},
-            n_samples=10,
+            n_samples=-1,
```
passed as expected

ricardoV94 replied:
You mean xfailed as expected?

matteo-pallini replied:
`test_binomial` was not expected to fail (as, it was not decorated with xfail), no?

ricardoV94 replied:
Nevermind I was confused by the collapsed diff

matteo-pallini committed:
Add missing n_samples for test_beta_binomial_selfconsistency

matteo-pallini committed:
Remove all the n_samples=-1 added for verifying tests robustness

matteo-pallini commented:
I removed all the `n_samples=-1` that have been added to make sure no xfail was missing and the ones we had were needed.

As far as I can tell, with the exception of the 3 points you listed above, everything else should be good to go. Let me know what you think

ricardoV94 committed:
Tweak normal_logcdf and moyal tests

Temporarily set n_samples=-1 to test all permutations.

ricardoV94 commented:
@drabbit17 Great work. I tweaked a bit the tests for the normal_cdf and moyal to see if the problems go away. Testing now with n_samples = -1. Fingers crossed!

ricardoV94 committed:
Reintroduce 32bit xfail on Moyal logcdf, revert n_samples and merge passing normal tests

ricardoV94 committed:
Small renaming and reordering

ricardoV94 committed:
Reintroduce accidentally removed n_samples from binomial test

ricardoV94 commented:
Okay this should be it. If nobody chimes in about the unclear xfail, I say we let it stay as a fossil. Your changes recoverd 3/4 of those tests so that's an improvement :)

Anything missing?

matteo-pallini commented:
It seems good to me. Thanks for assisting me while working the PR and reviewing it!

Out of curiosity how did you realize that moving the domain from `R` (which I assumes stands for real) to a custom one would have fixed the issue? specifically, how did you pick the custom one `Domain([-inf, -1.5, -1, -0.01, 0.0, 0.01, 1, 1.5, inf])`?

ricardoV94 commented:
> It seems good to me. Thanks for assisting me while working the PR and reviewing it!
> 
> Out of curiosity how did you realize that moving the domain from `R` (which I assumes stands for real) to a custom one would have fixed the issue? specifically, how did you pick the custom one `Domain([-inf, -1.5, -1, -0.01, 0.0, 0.01, 1, 1.5, inf])`?

I just tweaked the edge values of the R domain (the -inf, +inf are excluded). Since the issue was on the scipy side it was easy to test what was causing it to fail and whether the new values would fix it:

```python
import numpy as np
import scipy.stats as st                                                
                     
R = (-2.1, -1, -.01, 0, 0.01, 1, 2.1)                                   
Rplusbig = (0.5, 0.9, 0.99, 1, 1.5, 2, 20)     

for mu in R: 
    for sigma in Rplusbig: 
        for value in R: 
            if st.moyal(mu, sigma).logcdf(value) == -np.inf: 
                print(f'{mu=}, {sigma=}, {value=}') 
        
#  mu=2.1, sigma=0.5, value=-2.1
```
```python

R = (-2.1, -1, -.01, 0, 0.01, 1, 2.1)                                   
Rplusbig = (0.5, 0.9, 0.99, 1, 1.5, 2, 20)      
Custom = (-1.5, -1, -.01, 0, .01, 1, 1.5) 

for mu in R:  
     for sigma in Rplusbig:  
         for value in Custom:  
             if st.moyal(mu, sigma).logcdf(value) == -np.inf:  
                print(f'{mu=}, {sigma=}, {value=}')  

# Nothing is printed                                                       
```

michaelosthege commented on pymc3/tests/test_distributions.py:
```diff
@@ -902,6 +902,7 @@ def test_normal(self):
             R,
             {"mu": R, "sigma": Rplus},
             lambda value, mu, sigma: sp.norm.logcdf(value, mu, sigma),
+            decimal=select_by_precision(float64=6, float32=2),
```
`float32` precision of 2 digits sounds awfully bad. Are you sure about this?

ricardoV94 replied:
It's actually better than the normal::logp (this one is the normal::logcdf) test which uses a precision of 1! (Previous to this PR)

ricardoV94 replied:
This used to pass before because these tests were all seeded until very recently and the conditions with low precision were not appearing / failing since then

michaelosthege commented on pymc3/tests/test_distributions.py:
```diff
@@ -989,6 +971,27 @@ def test_wald(self, value, mu, lam, phi, alpha, logp):
+            Rplus,
+            {"mu": Rplus, "alpha": Rplus},
+            lambda value, mu, alpha: sp.invgauss.logpdf(value, mu=mu, loc=alpha),
+            decimal=select_by_precision(float64=6, float32=1),
```
Here too - 1 digit precision?!

ricardoV94 replied:
This preceeded this PR

ricardoV94 replied:
By the way these 1 precision comparisons are not necessarily as bad as they may sound. It's usually when the log becomes very negative, close to underflowing to -inf. things like -213.161 vs -213.172 that mean basically nothing.

Also many times we are dealing with loss of precision on the scipy side, not on our side.

michaelosthege reviewed (commented):
I see that you two already had a thorough discussion.

Just one thing: Instead of testing with a precision of 1 or 2, shouldn't we `xfail` these tests tighter precision on `float32`?
When the `xfail` stops failing we'll then know that the underlying numerics were improved and can tighten the test.

ricardoV94 commented:
I doubt we will ever notice xfails becoming xpasses specially because only a subset of combinations are tested in each run and xpasses already happen now and then by chance alone.

I feel that little precision is better than no checking at all (i.e. marking as xfail) provided this is only happening for the 32 bit version. If changes are done on our side that improve performance I think it's natural to try to increase precision in the tests when that happens.

ricardoV94 commented:
The case could be made to select testing points more carefully so that we don't test so often values that are numerically unstable in either our version or scipy's. 

That would deserve it's own PR, I think.

ricardoV94 commented:
Also, worth mentioning that it's really difficult to test float32 locally on a 64bit installation. For aesara we can change the config.floatX, but for the scipy counterpart I have not found a way to reliably emulate a float32 computation. Wrapping the output in a float32 does not suffice (in this PR I sometimes tried it locally and had the tests succeed with some precision, just to find they would still fail on the actual float32 CI runs here).

ricardoV94 commented:
For completeness here is a list of tests in the test_distributions.py module which are being evaluated with a precision of less than 3 decimals on float32:

* test_bound_normal::logp (float32=-1)
* test_truncated_normal::logp (float32=1)
* test_normal::logp (float32=-1)
* test_normal::logcdf (float32=2)  **# Changed in this PR**
* test_truncated_normal::logp (float32=1)
* test_half_normal::logp (float32=1)
* test_wald_custom_points::logp (float32=1)
* test_wald::logp (float32=1)
* test_mvnormal::logp (float32=-1 and float32=0 twice)
* test_matrixnormal::logp (float32=-1 and float32=0)
* test_ex_gaussian::logp (float32=2)
* test_ex_gaussian::logcdf (float32=2)
* test_logistic::logp (float32=1)
* test_logistic::logcdf (float32=1)
* test_logitnormal::logp (float32=1)

Additionally we have 14 tests marked as xfail on float32 

Total number of tests: 201

michaelosthege commented:
You're right that the precision doesn't deal with numbers before the comma.

Nevertheless we should make sure that `XPASS` tests show up as failures: https://docs.pytest.org/en/reorganize-docs/new-docs/user/xfail.html

ricardoV94 commented:
I didn't know about the strict mode to xfail. That looks useful.

I am afraid it would require some extra refactoring, because the tests (at least these we were looking at in `test_distributions.py`) don't always xfail, depending on which combination of parameters is tested. 

Should we open a new issue to address this possibility?

michaelosthege commented:
> Should we open a new issue to address this possibility?

Looking at the PR title I'd say it would be in the scope. But the PR has a lot of discussion & history already, so @drabbit17 @ricardoV94 you should decide how to proceed.
On their own, the current changes are also fine to merge, just don't fix the problem entirely.

ricardoV94 commented:
As I understood the goal of this PR was simply to "unshadow" tests that were unnecessarily bundled inside a shared xfail. This is done, I think.

What you are proposing is to now look at the xfails that remain relevant and see if we can make them strict so that xPasses would trigger them. There are some tests that look deterministic so in theory they could, but I would guess the majority cannot be made strict, unless we change the actual tests as they fail randomly. This would require looking at all the xfails, not just the ones that were changed in this PR due to "bundling".

What do you think @drabbit17?

PS: I agree this would be valuable!

matteo-pallini commented:
Thanks for suggesting introducing the strict mode for the test. I also think that using that option would be valuable.

I think having that change in another PR would make the process tidier, as mentioned we already have a lot of discussion in this PR. 

Also, I think that in the original scope of this PR we were only meant to extract the non-failing sub-checks from the xfail decorated ones and not change the behavior of any individual test/check. Although we ended up changing by modifying the support of some tests and the check precision (and possibly some other stuff I cannot think of). Adding the strict fail would add further changes to how the tests behave. 

Having said that, I feel that these are mostly philosophical considerations :-). In practice, I don't have a strong opinion on whether to add the strict change in this PR or in another one. Happy to do it however you guys feel it's better (or to flip a coin to decide). I also agree that's a change worth adding.

michaelosthege reviewed (approved):
Let's merge

ricardoV94 commented:
Thanks @drabbit17! Looking forward to your next PR
